README.md
TurboBench: Compressor Benchmark[![Build Status](https://travis-ci.org/powturbo/TurboBench.svg?branch=master)](https://travis-ci.org/powturbo/TurboBench)
================================
+ **TurboBench**
 - The only benchmark program including **[LzTurbo](https://sites.google.com/site/powturbo/)**
 - 100% in-memory benchmark, no I/O overhead
 - Include (>60) allmost all popular, latest or fastest compressors in one compiled package
 - No other compressor benchmark includes more codecs or offer more precision and features
 - Benchmarking **Entropy Coders**, **Lz77**, **Rolz**, **BWT** and **Context mixing** compressors
 - **Minimum** plugins call **overhead**
 - Set one, a group or several compressors to benchmark at the command line
 - **Multiple** input files with **recursive** directories
 - Concatenate **multiple small files** into one multiblock file
 - Benchmark **multiblock** file as one large block, but each block processed separatelly
 - Avoid **cache szenario** found in other benchmarks with small files
 - Set block size, file size limit,...
 - Set number of **iterations**, number of **runs**, **benchmarks**, set max. **time** per run
   and for all benchmarks.
 - Automatic **sort**
 - :new: automatic update & merge of result files
 - :new: **Text**, **html**, **csv**, **markdown** and other output formats without retesting
 - :new: html output with sortable tables
 - :new: **Transfer speed sheet** for different connections or devices: GPRS,2G,3G,4G,DSL,Network,HDD,SSD,RAM
 - :+1: **Html plot:** **Speedup** + **Speed/Ratio**
 - **Linux** and **Windows** binaries
 - 100% C/C++, w/o inline assembly 
 - Enable/disable groups or individual codecs at compile time
 - All in one executable, no hassless installing of additional packages, graphic libraries, python,...
 - :+1: build in peak memory usage reporting for compression and decompression in html output
 - :new: **Encoding** and **Transform** plugins 

### Download benchmark executable incl. LzTurbo:
 - [TurboBench Linux](https://sites.google.com/site/powturbo/downloads)
 - [TurboBench Windows](https://sites.google.com/site/powturbo/downloads)

### Benchmark:
CPU: Sandy bridge i7-2600k at 4.2GHz, gcc 5.2, ubuntu 15.10, single thread.
- Realistic and practical benchmark with large files
- No PURE cache benchmark

##### - Data files:
#### TurboBench compressor benchmark:
- File [app3.tar binary Portable Apps Suite](http://compressionratings.com/download.html)

- [html output+Speedup](http://htmlpreview.github.io/?https://github.com/powturbo/TurboBench/master/turbobench_/app3.tar.html)
  (download app3.tar.html to local file for better browsing)

------------------------------------------------------------------------
![Seedup](turbobench_/app3.tar.png "Speedup: Transfer+decompression")
------------------------------------------------------------------------
![Speed/Ratio](turbobench_/app3r.tar.png "Speed/Ratio: Decompression")
------------------------------------------------------------------------

 (bold = pareto)  MB=1.000.000

|C Size|ratio%|C MB/s|D MB/s|Name|
|--------:|-----:|--------:|--------:|----------------|
|32798929| 32.8|**2.79**|**65.49**|**lzma 9**|
|32922377| 32.9|1.61|**69.65**|**lzturbo 49**|
|33761620| 33.7|2.64|**277.04**|**lzham 4**|
|34109576| 34.1|2.17|**1318.56**|**lzturbo 39**|
|35638896| 35.6|1.19|950.41|zstd 20|
|36944243| 36.9|**69.99**|**1411.77**|**lzturbo 32**|
|37313258| 37.3|2.40|**2149.57**|**lzturbo 29**|
|41668560| 41.6|0.19|246.43|brotli 11|
|45799999| 45.8|27.03|354.38|brotli 6|
|46304388| 46.3|35.75|347.99|brotli 5|
|46492542| 46.4|**191.69**|1186.68|**lzturbo 31**|
|46875269| 46.8|48.58|695.99|zstd 9|
|48836109| 48.8|113.95|676.74|zstd 5|
|49324183| 49.3|141.06|337.46|brotli 1|
|49860700| 49.8|16.94|295.99|zlib 9|
|49918739| 49.9|**288.52**|1082.09|**lzturbo 30a**|
|49962678| 49.9|35.70|294.24|zlib 6|
|50027825| 50.0|52.85|1899.72|lzturbo 22|
|50311200| 50.3|**312.35**|1090.91|**lzturbo 30**|
|50337788| 50.3|6.73|1428.58|lz5 9|
|52597358| 52.5|262.30|2068.57|lzturbo 21|
|52928477| 52.9|69.17|276.75|zlib 1|
|53112430| 53.1|298.70|442.42|zstd 1|
|54265487| 54.2|2.01|**3883.96**|**lzturbo 19**|
|54423519| 54.4|46.33|1918.40|lz4 9|
|55400947| 55.3|**465.29**|1900.38|**lzturbo 20a**|
|55764172| 55.7|405.79|1517.45|lz5 1|
|55923645| 55.9|141.96|3713.00|lzturbo 12|
|57606731| 57.6|263.89|3500.06|lzturbo 11|
|59090242| 59.0|**631.43**|2099.72|**lzturbo 20**|
|60016380| 60.0|596.92|3361.22|lzturbo 10a|
|61460109| 61.4|**705.32**|3500.41|**lzturbo 10**|
|61938605| 61.9|671.21|2069.81|lz4 1|
|100098564|100.0|**8647.84**|**8408.10**|**memcpy**|

### Plugins:
#### Compressor Lz77,Rolz,Bwt,zpaq:
 - [LzTurbo v1.3](https://sites.google.com/site/powturbo)
 - [balz v1.20](http://sourceforge.net/projects/balz) 
 - [bcm v1.0](http://sourceforge.net/projects/bcm) 
 - [Blosc v2.0](https://github.com/Blosc/c-blosc2)
 - [BriefLz v1.1.0](https://github.com/jibsen/brieflz) 
 - [Brotli v16-01](https://github.com/google/brotli) 
 - [Bzip2 v1.06](http://www.bzip.org/downloads.html) 
 - [Chameleon v15-03](http://cbloomrants.blogspot.de/2015/03/03-25-15-my-chameleon.html) 
 - [Crush v1.0.0](http://sourceforge.net/projects/crush) 
 - [CSC v16-01](https://github.com/fusiyuan2010/CSC) 
 - [Density v0.13.0](https://github.com/centaurean/density) 
 - [Doboz v14-01-14](https://bitbucket.org/attila_afra) 
 - [FastLz v0.1.0](http://fastlz.org) 
 - [Gipfeli v15.12](https://github.com/google/gipfeli) 
 - [heatshrink v0.4.1](https://github.com/atomicobject/heatshrink) 
 - [bsc v3.1.0](https://github.com/IlyaGrebnov/libbsc) 
 - [LibLZF v1.06](http://oldhome.schmorp.de/marc/liblzf.html) 
 - [LibLzg v1.0.8](https://github.com/mbitsnbites/liblzg) 
 - [Lz4 v1.7.1](https://github.com/Cyan4973/lz4) 
 - [Lz5 v1.3.3](https://github.com/inikep/lz5) 
 - [Lzfse](https://developer.apple.com/library/mac/samplecode/CompressionSample/Introduction/Intro.html) *
 - [Lzham v1.1/15-12](https://github.com/richgel999/lzham_codec_devel) 
 - [Lzlib v1.7](http://www.nongnu.org/lzip) 
 - [Lzmat v1.0](https://github.com/nemequ/lzmat) 
 - [Lzma v9.35](http://7-zip.org) 
 - [Lzo v2.09](http://www.oberhumer.com/opensource/lzo) 
 - [Miniz v13-10](https://github.com/richgel999/miniz) :new:
 - [ms-compress v15.09](https://github.com/coderforlife/ms-compress) 
 - [Nakamichi v15-10](http://www.overclock.net/t/1577282/fastest-open-source-decompressors-benchmark#post_24538188) :new:
 - [Pithy v2011](https://github.com/johnezang/pithy) 
 - [Quicklz v1.5.1](http://www.quicklz.com) 
 - [sap v15-11](https://github.com/CoreSecurity/pysap) 
 - [shoco v2015](https://github.com/Ed-von-Schleck/shoco) 
 - [Shrinker v0.1/r9](https://code.google.com/p/data-shrinker) 
 - [Snappy v1.1.3](https://github.com/google/snappy) 
 - [Snappy-c v1.1.2/14.04](https://github.com/andikleen/snappy-c) 
 - [Tornado v0.6a](http://freearc.org) 
 - [wfLZ v15-04](https://github.com/ShaneWF/wflz) 
 - [Yalz77 v15-09](https://github.com/ivan-tkatchev/yalz77) 
 - [Yappy v2011]() 
 - [zlib v1.2.8](http://zlib.net)
 - [Libzling v16-01](https://github.com/richox/libzling) 
 - [Zopfli v15-05](https://code.google.com/p/zopfli) 
 - [ZSTD v0.4.6](https://github.com/Cyan4973/zstd) 
 - [zpaq v7.05](https://github.com/zpaq/zpaq) 

* LzFSE: included but not tested

#### Entropy coder:
###### bitwise range coder
 - [TurboRC-Range Coder v1.3](https://sites.google.com/site/powturbo)
 - [Bitwise RC v2010](http://encode.ru/threads/1153-Simple-binary-rangecoder-demo)
 - [Bitwise vector RC v2012](http://encode.ru/threads/1200-Vectorized-rangecoder)
 - [bcm range coder v1.0](http://sourceforge.net/projects/bcm)
 - [FastAri v15-10](https://github.com/davidcatt/FastARI)

###### bytewise range coder
 - [TurboAC v1.3](https://sites.google.com/site/powturbo)
 - [subotin range coder v2000](http://ezcodesample.com/ralpha/Subbotin.txt)
 - [Fast AC v2006](http://www.cipr.rpi.edu/research/SPIHT/)
 - [Range Coder/J.Bonfield v15-07](ftp://ftp.sanger.ac.uk/pub/users/jkb)
 - [FQZ/PPMD Range Coder v15-03](http://encode.ru/threads/2149-ao0ec-Bytewise-adaptive-order-0-entropy-coder)
 - [PPMD Range Coder v15-03](http://encode.ru/threads/2149-ao0ec-Bytewise-adaptive-order-0-entropy-coder)

###### ABS: Asymmetric binary systems 
 - [Fpaqc:Asymmetric Binary Coder v07-12](http://www.mattmahoney.net/dc/)

###### ANS: Asymmetric Numeral Systems
 - [TurboANX-ANS v1.3](https://sites.google.com/site/powturbo)
 - [Finite State Coder v15-05](https://github.com/skal65535/fsc)
 - [Finite State Entropy v15-11](https://github.com/Cyan4973/FiniteStateEntropy)
 - [ANS/J.Bonfield v15-08](ftp://ftp.sanger.ac.uk/pub/users/jkb)
 - [Nania Adaptive rANS v2015](http://encode.ru/threads/2079-nARANS-(Nania-Adaptive-Range-Variant-of-ANS))

###### Huffman Coding
 - [TurboHF-Huffmann v1.3](https://sites.google.com/site/powturbo)
 - [Tornado Huf v0.6a](http://freearc.org/Research.aspx) 
 - [zlib Huffmann v1.2.8](https://github.com/Cyan4973/FiniteStateEntropy) 
 - [Fast HF v2006](http://www.cipr.rpi.edu/research/SPIHT/) 
 - [Polar Codes v10-07](http://www.ezcodesample.com/prefixer/prefixer_article.html)
 - [inline memcpy](https://github.com/powturbo/TurboBench)
 - [library memcpy](https://github.com/powturbo/TurboBench)

#### Encoding:
 - [TurboRLE](https://github.com/powturbo/TurboRLE)

#### Transform:
 - [bwt:libdivsufsort](https://github.com/y-256/libdivsufsort)

### Testing:
  + List all plugins: "./turbobench -l2"<br />
  + List all compiled codecs: "./turbobench -l1"<br />
  + type "./turbobench -h" for help

##### - Groups FASTEST,FAST,EFFICIENT,MAX,OPTIMAL,BWT:
  + test all fast compressors in the lz4, lzturbo, zlib class<br />


        ./turbobench -eFAST file

##### - Codecs:

  + individual codec test (output to screen & file.tbb)<br />


        ./turbobench -elzturbo,19,29,39/brotli,6/zlib,6 file

  + retest or test other compressors and merge the results to file.tbb<br />


        ./turbobench -eFAST/bzip2 file

##### - Print + Plot

   + Print result file + "transfer+decompression speedup" plot to file.html for browsing


        ./turbobench -p2 -S2 file.tbb

   
### Compile:

  		git clone --recursive git://github.com/powturbo/TurboBench.git
        cd TurboBench
  		make

### Environment:
###### OS/Compiler (32 + 64 bits):
- Linux: GNU GCC (>=4.6)
- clang (>=3.2) 
- Windows: MinGW
- Windows: Visual Studio 2015

### References:
- [Speedup](http://cbloomrants.blogspot.de/2015/03/03-02-15-oodle-lz-pareto-frontier.html) - "speedup" is the ratio of the effective speed vs. the speed of the disk
- [CompFuzz Results](https://github.com/nemequ/compfuzz/wiki/Results) - list of vulnerable codecs
- [Lzbench](https://github.com/inikep/lzbench) - lzbench is an in-memory benchmark of open-source compressors

Last update: 15 JAN 2016



lzlib/NEWS
Changes in version 1.7:

The fast encoder, which produces a compression speed and ratio
comparable to those of gzip, has been ported from lzip.

The option "-0" has been ported from lzip to minilzip.

If all the data to be compressed are written in advance, lzlib will
automatically adjust the header of the compressed data to use the
smallest possible dictionary size. This feature reduces the amount of
memory needed for decompression and allows minilzip to produce identical
compressed output as lzip.

The targets "install-compress", "install-strip-compress",
"install-info-compress" and "install-man-compress" have been added to
the Makefile.


plugreg.c
/**
    Copyright (C) powturbo 2013-2015
    GPL v2 License
  
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

    - homepage : https://sites.google.com/site/powturbo/
    - github   : https://github.com/powturbo
    - twitter  : https://twitter.com/powturbo
    - email    : powturbo [_AT_] gmail [_DOT_] com
**/
//	    TurboBench: plugreg.c - registry 
#define C_EC   0x0001 
#define C_EC1  0x0002 

#define E_FLAG 0x0f00
#define E_ANS  0x0100
#define E_HUF  0x0200

struct plugs { int id; char *s,*v,*lev; unsigned flag,blksize; };
  
struct plugs plugs[] = {
    #ifdef _PLUGIN2
  #include "../beplugr.h"
    #endif

    #ifdef _BALZ
  { P_BALZ, 	"balz", 			_BALZ, "0,1" }, 
    #endif
	
    #ifdef _BCM
  { P_BCM, 		"bcm", 				_BCM, "" }, 
    #endif

    #ifdef _C_BLOSC2
  { P_C_BLOSC2, "blosc",			_C_BLOSC2, "0,1,2,3,4,5,6,7,8,9", 64*1024},
    #endif
	   
    #ifdef _BRIEFLZ
  { P_BRIEFLZ,	"brieflz", 		    _BRIEFLZ, "" }, 
    #endif
	
    #ifdef _BROTLI
  { P_BROTLI,	"brotli", 			_BROTLI, "0,1,2,3,4,5,6,7,8,9,11/DOWX" },
    #endif
	 
    #ifdef _BZIP2
  { P_BZIP2,	"bzip2", 			_BZIP2, "" }, 
    #endif

    #ifdef _CHAMELEON
  { P_CHAMELEON,"chameleon",		_CHAMELEON, "1,2" },
    #endif

    #ifdef _CRUSH
  { P_CRUSH,	"crush", 			_CRUSH, "0,1,2" },
    #endif
	
    #ifdef _CSC
  { P_CSC,	    "csc", 				_CSC, "1,2,3,4,5" },
    #endif
	
    #ifdef _DENSITY
  { P_DENSITY, 	"density",        	_DENSITY,"1,2,3" },
    #endif

    #ifdef _DOBOZ
  { P_DOBOZ,	"doboz",			_DOBOZ, "" },  //crash on windows
    #endif

    #ifdef _FASTLZ
  { P_FASTLZ,	"fastlz", 			_FASTLZ,"1,2" },
    #endif
	
	#ifdef _GIPFELI
  { P_GIPFELI, 	"gipfeli", 			_GIPFELI, "" }, 
    #endif
    
    #ifdef _HEATSHRINK 
  { P_HEATSHRINK,"heatshrink",		_HEATSHRINK, "" },
    #endif
	
    #ifdef _LIBBSC
  { P_LIBBSC_ST,"bsc_st", 			_LIBBSC, "3,4,5,6,7,8" }, 
  { P_LIBBSC, 	"bsc", 				_LIBBSC, "1,2"}, 
    #endif

    #ifdef _LIBLZF
  { P_LIBLZF, 	"lzf", 				_LIBLZF, "" },
    #endif

    #ifdef _LIBLZG
  { P_LIBLZG,  	"lzg", 				_LIBLZG, "1,2,3,4,5,6,7,8,9" }, 
    #endif
	 
    #ifdef _LIBZPAQ
  { P_LIBZPAQ,  "zpaq", 			_LIBZPAQ, "0,1,2,3,4,5" }, 
    #endif

    #ifdef _LZ4
  { P_LZ4,  	"lz4",				_LZ4, "0,1,9" }, 
    #endif
	
    #ifdef _LZ5
  { P_LZ5,  	"lz5",				_LZ5, "0,1,9" }, 
    #endif
	
    #ifdef _LZFSE
  { P_LZFSE, 	"lzfse", 			_LZFSE, "" },
    #endif

    #ifdef _LZHAM
  { P_LZHAM, 	"lzham", 			_LZHAM,"1,2,3,4/x" }, 
    #endif

    #ifdef _LZLIB 
  { P_LZLIB, 	"lzlib", 			_LZLIB, "1,2,3,4,5,6,7,8,9" },
   #endif
	
    #ifdef _LZMAT
  { P_LZMAT, 	"lzmat", 			_LZMAT, "" },
    #endif
  
    #ifdef _LZMA
  { P_LZMA,  	"lzma", 			_LZMA, "0,1,2,3,4,5,6,7,8,9" }, 
    #endif

    #ifdef _LZO
	  #ifdef _MINILZO
  { P_LZO1x, 	"lzo1x",			_LZO, "1" },
      #else
  { P_LZO1b, 	"lzo1b", 			_LZO, "1,9,99,999" },  
  { P_LZO1c, 	"lzo1c",			_LZO, "1,9,99,999" },
  { P_LZO1f, 	"lzo1f",			_LZO, "1,999" },
  { P_LZO1x, 	"lzo1x", 			_LZO, "1,11,12,15,999" },
  { P_LZO1y,  	"lzo1y",			_LZO, "1,999" }, 
  { P_LZO1z, 	"lzo1z", 			_LZO, "999" }, 
  { P_LZO2a, 	"lzo2a", 			_LZO, "999" }, 
      #endif
    #endif
   
    #ifdef _LZSS
  { P_LZSS,  	"lzss", 			_LZSS, "1,2" }, 
    #endif

    #ifdef _LZTURBO
  { P_LZT,		"lzturbo", 			_LZTURBO, "10,11,12,19,20,21,22,29,30,31,32,39,49/a" },
    #endif

    #ifdef _MINIZIP
  { P_MINZ,    	"miniz", 			_MINIZIP, "1,2,3" }, 
    #endif

    #ifdef _MSCOMPRESS
  { P_MSCOMPRESS,"mscompress", 		_MSCOMPRESS, "2,3,4" }, 
    #endif

    #ifdef _PITHY
  { P_PITHY, 	"pithy",			_PITHY, "0,1,2,3,4,5,6,7,8,9" },
    #endif

    #ifdef _QUICKLZ
  { P_QUICKLZ, 	"quicklz",			_QUICKLZ, "1,2,3" },
    #endif

    #ifdef _SAP
  { P_SAP, 	    "sap",				_SAP, "0,1,2"	},
    #endif
  
    #ifdef _SHRINKER
  { P_SHRINKER, "shrinker",			_SHRINKER, "", 0, (1<<26) },
    #endif

    #ifdef _SHOCO
  { P_SHOCO,    "shoco",			_SHOCO, "" },
    #endif

    #ifdef _SNAPPY
  { P_SNAPPY, 	"snappy",			_SNAPPY, ""	},
    #endif
  
    #ifdef _SNAPPY_C
  { P_SNAPPYC, 	"snappy_c",			_SNAPPY_C,"" },
    #endif
   
    #ifdef _TORNADO
  { P_TORNADO, 	"tornado", 			_TORNADO, "1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16" }, 
    #endif
   
    #ifdef _WFLZ
  { P_WFLZ,	    "wflz", 			_WFLZ, "1,2" },
    #endif
	
    #ifdef _WIMLIB
  { P_WIMLIB,	"wimlib", 			_WIMLIB, "1,2,3" },//LZX,XPRESS,LZMS
    #endif
	
    #ifdef _WKDM
  { P_WKDM, 	"WKdm",				_WKDM, "" }, // crash
    #endif	

    #ifdef _YALZ77
  { P_YALZ77, 	"yalz77", 			_YALZ77, "1,6,12" },
    #endif

    #ifdef _YAPPY
  { P_YAPPY, 	"yappy",			_YAPPY, "" },//crash windows
    #endif

    #ifdef _ZLIB
  { P_ZLIB, 	"zlib", 			_ZLIB, "1,2,3,4,5,6,7,8,9" },
    #endif 			
  
    #ifdef _ZLING
  { P_ZLING, 	"zling", 	   		_ZLING, "0,1,2,3,4" }, 
    #endif

	#ifdef _ZOPFLI
  { P_ZOPFLI, 	"zopfli",			_ZOPFLI, ""}, 
    #endif
	
    #ifdef _ZSTD
  { P_ZSTD, 	"zstd", 			_ZSTD, "1,2,3,4,5,6,7,8,9,15,16,20" },
    #endif
//-----------------------------------------------------------------------------------	  
  { P_MCPY, 	"imemcpy", 			_IMEMCPY, "" },
  { P_LMCPY, 	"memcpy",			_MEMCPY,  "" },

    #ifdef _LZTANS
	   #ifdef __x86_64__
  { P_LZTANS64, "TurboANS",			_LZTANS, "" },
       #endif
  { P_LZTANS, 	"TurboANX",	    	_LZTANS, "",C_EC },
//{ P_LZTANSX, 	"TurboANXX",	  	_LZTANS, "",C_EC1},
    #endif
	
    #ifdef _LZTHF
  { P_LZT3, 	"TurboHF",		    _LZTHF, "",C_EC},  
    #endif
    #ifdef _LZTRC
  { P_LZT5, 	"TurboAC_byte", 	_LZTRC, "",C_EC}, 
  { P_LZT6, 	"TurboRC", 			_LZTRC, "",C_EC}, 
  { P_LZT7, 	"TurboRC_o1",   	_LZTRC, "",C_EC1}, 
    #endif

    #ifdef _BCMEC
  { P_BCMEC, 	"bcmec", 			_BCMEC, "", C_EC},
    #endif

    #ifdef _FSC
  { P_FSC, 		"fsc", 				_FSC, "", C_EC|E_ANS },
    #endif

    #ifdef _FSE
  { P_FSE, 		"fse", 				_FSE, "", C_EC|E_ANS },
  { P_FSEH,		"fsehuf", 			_FSE, "", C_EC|E_HUF },
    #endif

    #ifdef _MMABS
  { P_MMABS, 	"fpaqc",	 		_MMABS, "", C_EC},
    #endif

    #ifdef _FASTHF
  { P_FASTHF,  	"FastHF", 			_FASTHF, "",C_EC },
    #endif

    #ifdef _FASTARI
  { P_FASTARI,  "FastAri", 			_FASTARI, "",C_EC },
    #endif

    #ifdef _FASTAC
  { P_FASTAC,  	"FastAC", 			_FASTAC, "",C_EC },
    #endif

    #ifdef _JAC
  { P_JAC, 		"arith_static",		_JAC, "",C_EC|E_ANS },
    #endif
  
	#ifdef _JRANS
  { P_JRANS, 	"rans_static4c",	_JRANS, "",C_EC |E_ANS},
  { P_JRANS1, 	"rans_static4c_o1", _JRANS, "",C_EC1|E_ANS},
  { P_JRANS64, 	"rans_static64c",	_JRANS, "",C_EC |E_ANS},
  { P_JRANS641,	"rans_static64c_o1",_JRANS, "",C_EC1|E_ANS},
    #endif
 
    #ifdef _MMRC
  { P_FPAQ0F2, 	"fpaq0f2", 		    _MMRC, "",C_EC},
  { P_MMRC, 	"fpaq0p_mm", 		_MMRC, "",C_EC},
  { P_O0ZP,     "o0zp",             _MMRC, "",C_EC},
    #endif

	#ifdef _NANS
  { P_NANS, 	"naniarans",		_NANS, "",C_EC|E_ANS},
    #endif
 
	#ifdef _POLHF
  { P_POLHF,    "polar", 			_POLHF, "",C_EC},
    #endif

    #ifdef _SUBOTIN
  { P_SUB, 		"subotin", 			_SUBOTIN, "",C_EC},
    #endif

    #ifdef _TORNADOHF
  { P_TORNADOHF,"tornado_huff", 	_TORNADOHF, "", C_EC},
    #endif

    #ifdef _ZLIBH
  { P_ZLIBH, 	"zlibh",			_ZLIBH, "", C_EC|E_HUF },
    #endif
	
  { -1, "", ""} 
  };
